### DISCUSSION TITLE
PostgresSyntaxError when trying asyncpg and  text(...).bindparams(...)

### DISCUSSION THREAD

**Discussion by 0xD09CD196D188D0B0** [sqlalchemy/sqlalchemy#7192] [Usage Questions]
Hello!

If you execute a simple parameterized query using AsyncSession, like:
```
# db_session is instance of AsyncSession
stmt = text('SET local statement_timeout = :value').bindparams(value=500)
await db_session.execute(stmt)
```

you end up with:
```
E                   sqlalchemy.exc.ProgrammingError: (sqlalchemy.dialects.postgresql.asyncpg.ProgrammingError) <class 'asyncpg.exceptions.PostgresSyntaxError'>: syntax error at or near "$1"
E                   [SQL: SET local statement_timeout = %s]
E                   [parameters: (500,)]
E                   (Background on this error at: https://sqlalche.me/e/14/f405)
```

But it perfectly works if you just use an f-string, but using f-string does not sound cool.

I've found some explanations from asyncpg side (https://github.com/MagicStack/asyncpg/issues/605#issuecomment-683192217):

> asyncpg does not do query argument interpolation in any form. The query you write gets passed directly to PostgreSQL. This is unlike psycopg2 which has extensive query rewriting/interpolation mechanisms.

Is it maybe fixable on SQLAlchemy side? Like SQLAlchemy passes asyncpg queries with parameter values already placed.

**Comment by CaselIT** [comment]
Hi,

The same error happens if you run the same query with asyncpg:
```py
>>> async def go():
...   conn = await asyncpg.connect('postgresql://scott:tiger@localhost:5432/test')
...   await conn.execute('SET local statement_timeout = $1', 1000)

>>> asyncio.run(go())
Traceback (most recent call last):
  ...
  File "asyncpg\protocol\protocol.pyx", line 168, in prepare
asyncpg.exceptions.PostgresSyntaxError: syntax error at or near "$1"
```

The issue here seems of asyncpg and in part of postgresql that does not accept parameters in that query. My suggestion is to ask for advice in the asyncpg repository.

> Is it maybe fixable on SQLAlchemy side? Like SQLAlchemy passes asyncpg queries with parameter values already placed.

No, this is not something SQLAlchemy does, and in general not something we are interested in doing, since it has a lot of security implications

**Reply by CaselIT** [reply]
Actually, you can use this way to bind the parameters. https://docs.sqlalchemy.org/en/14/faq/sqlexpressions.html#rendering-bound-parameters-inline

I though it did not apply to `text` but it seems it does.

In any case I would not trust user input directly to be passed there. Since the parameter is an integer, probably doing something like `int(input)` would be ok-ish

**Reply by 0xD09CD196D188D0B0** [reply]
Thanks for suggestions.

> My suggestion is to ask for advice in the asyncpg repository.

Somebody already asked some time ago -> https://github.com/MagicStack/asyncpg/issues/605

and unfortunately their answer is:

> This is outside of scope and can be solved by a wrapper library.

**Reply by CaselIT** [reply]
> Somebody already asked some time ago -> [MagicStack/asyncpg#605](https://github.com/MagicStack/asyncpg/issues/605)

It was for a different case ,a create table compared to a set, but I guess the answer should be similar also for this query.

**Reply by Alex-Bogdanov** [reply]
> Hi,
> 
> The same error happens if you run the same query with asyncpg:
> 
> ```python
> >>> async def go():
> ...   conn = await asyncpg.connect('postgresql://scott:tiger@localhost:5432/test')
> ...   await conn.execute('SET local statement_timeout = $1', 1000)
> 
> >>> asyncio.run(go())
> Traceback (most recent call last):
>   ...
>   File "asyncpg\protocol\protocol.pyx", line 168, in prepare
> asyncpg.exceptions.PostgresSyntaxError: syntax error at or near "$1"
> ```
> 
> The issue here seems of asyncpg and in part of postgresql that does not accept parameters in that query. My suggestion is to ask for advice in the asyncpg repository.
> 
> > Is it maybe fixable on SQLAlchemy side? Like SQLAlchemy passes asyncpg queries with parameter values already placed.
> 
> No, this is not something SQLAlchemy does, and in general not something we are interested in doing, since it has a lot of security implications

@CaselIT actually it's a kinda weird, but the latest version of `asyncpg==0.30.0` already support passing positional arguments when executing the raw SQL query, see the ref [doc](https://magicstack.github.io/asyncpg/current/usage.html#asyncpg-usage). So, I still can't get the idea what's a point of not implementing this feature in SQLAlchemy, because in version `2.0.36` this issue mentioned by @mvbrn is still reproducible

**Reply by CaselIT** [reply]
I'm not sure what you expect sqlalchemy to implement here. can you clarify?
